### PR TITLE
Submit Evaluator task to AzureBatch.

### PR DESCRIPTION
**Pull request by homezcx**
Add Runtime Start/Stop Handler and Resource Release/Request/Launch Handler. Submit Evaluator task to AzureBatch.

This patch is going submit the evaluator task to Azure Batch from Driver Runtime. "Hello Reef" will be printed in stdout.txt in evaluator task node.

Please note:
1. AzureBatch credential is still hard coded in DriverConfiguration. This should be eventually provided by client code. Mentioned in TODO.

2. Evaluator Java commandline is hard coded now. It should be configurable. 

3. We need to figure out how to pass through Job Id, so we don't need to create a new Job for submitting evaluator task.

4. We need to investigate "NodeDescriptorHandler" use for evaluator. As Batch concept, this node is not needed but it is required by evaluator. TODO: replace the dummy setup.

### CODE REVIEW

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchDriverConfiguration.java
Comment (ryumbra) [inline lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchDriverConfiguration.java:35]: FYI: on Friday I checked in a change that adds these parameters to azure batch runtime. I also completely removed hdinisght runtime dependency from batch runtime. So you may want to merge the changes.

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchDriverConfiguration.java
Comment (ryumbra) [inline lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchDriverConfiguration.java:112]: What's the difference between AzureBatchRuntimeStartHandler and AzureBatchResourceManagerStartHandler?

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceRequestHandler.java
Comment (ryumbra) [inline lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceRequestHandler.java]: NIT: extra line.

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java
Comment (ryumbra) [inline lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java]: NIT: lowercase "L".

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java
Comment (ryumbra) [inline lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java]: This shouldn't be a job manager task.

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java
Comment (ryumbra) [inline lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java]: Actually, you can find the ID of the job by looking at "AZ_BATCH_JOB_ID" environment variable. Or, you can pass it via Driver config. Either approach is fine with me.
https://docs.microsoft.com/en-us/azure/batch/batch-compute-node-environment-variables

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java
Comment (ryumbra) [inline lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java]: I suggest we use the JOB ID environment variable instead of creating a new job.

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchDriverConfiguration.java
Comment (homezcx) [reply lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchDriverConfiguration.java:112]: The base class isResourceManagerStartHandler not RuntimeStartHandler. So I renamed it. Is there any specific reason to name it as RuntimeStartHandler?

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java
Comment (homezcx) [reply lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java]: mentioned in Note. This will be implemented in follow up task, if we all agree using environment variable or other approach.

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java
Comment (homezcx) [reply lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java]: mentioned in Note and this function as TODO. This will be implemented in follow up task.

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java
Comment (homezcx) [reply lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java]: You are right. I mentioned in the node and code TODO comment.

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchDriverConfiguration.java
Comment (homezcx) [reply lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchDriverConfiguration.java:35]: I will rebase it before merging the change.

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceRequestHandler.java
Comment (homezcx) [reply lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceRequestHandler.java]: will fix this.

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java
Comment (ryumbra) [reply lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java]: If nothing else, I'd suggest this to be one thing you fix before checking in. Getting the job ID from the environment variable will be one line change and adding a job to a task will be easier than creating a new job. Plus, it will unblock me for the rest of the day :)
(Otherwise, I'll have to do once I pull your change)

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchDriverConfiguration.java
Comment (ryumbra) [reply lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchDriverConfiguration.java:112]: Got it. Just seemed strange choice of naming.

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java
Comment (homezcx) [reply lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java]: Sure. Not a problem.

Comment (ryumbra) [reply lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java]: Thank you!

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java
Comment (ryumbra) [inline lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java]: NIT: evaluator task.

**Review by ryumbra — APPROVED** [review]

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchDriverConfiguration.java
Comment (homezcx) [reply lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchDriverConfiguration.java:35]: done

File: lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java
Comment (homezcx) [reply lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java]: Fixed.

**Pull request merged**
